### DISCUSSION TITLE
modem_cellular: ability to launch chat script from application space.

### DISCUSSION THREAD

**Discussion by RomainPelletant** [zephyrproject-rtos/zephyr#64293] [Q&A] [answered]
Is there a way to execute a chat script directly on a modem interface based on modem cellular?

For example, sometimes I would initiate a QCSQ command to get network health status, from application space to a specific device and I didn't find a better way than add a specific command reachable from application space:
[011_cellular_add_run_cmds.patch](https://github.com/zephyrproject-rtos/zephyr/files/13112908/011_cellular_add_run_cmds.patch)

Do you have a better idea?

**Comment by RomainPelletant** [comment]
@bjarki-trackunit It could require a generic function to get information on network signals.
If I correctly understand modem_cellular driver, we shall implement a specific chat script for each modem's models.
To avoid direct calls to the driver, a specific API shall be provided to request to commands like 

```
net_mgmt(NET_REQUEST_WIFI_IFACE_STATUS, iface, &status,
		     sizeof(struct wifi_iface_status))
```
I don't know if net_mgmt is the right place for that or a new one shall be implemented (more cellular network specific)

Data could be:

- SINR
- RSRP
- RSRQ
- LAC
- CellID
- RSSI
- Operator code

Maybe a RFC could be more appropriate for that but I would have your opinion.
Thanks for your time :)

**Reply by bjarki-andreasen** [reply, marked as answer]
The modem_cellular driver is an MVP intended as a base implementation for other drivers. It demonstrates how the modem modules work, and provides network connectivity.

We do not have a proper cellular modem API in zephyr yet, so custom APIs must be implemented out-of-tree for now. this includes setting APN and other credentials, selecting radio access tech, sending sms'es etc. In short, copy paste modem_cellular.c out of tree, create an API that fits your needs out-of-tree, then implement it in your out-of-tree driver

This is taken from the modem_cellular driver line 1487:
```
DEVICE_DT_INST_DEFINE(inst, modem_cellular_init, PM_DEVICE_DT_INST_GET(inst),		\
			    &MODEM_CELLULAR_INST_NAME(data, inst),				\
			    &MODEM_CELLULAR_INST_NAME(config, inst), POST_KERNEL, 99, <your api here>);
```

This way you get full flexibility to implement exactly what you need using the device driver model. This is the intended purpose of the modem modules.

**Reply by RomainPelletant** [reply]
Thanks @bjarki-trackunit  :)

**Reply by AndrewAWG97** [reply]
Hi @bjarki-trackunit, I want to use the same way you're using to communicate with the modem with scripts. I want to get a GPS signal. 
I tried to add commands to the script, they work but as you know GPS signal may need to send command many times until it return location, but in our case we can't send the same command if it returns an error. 
I thought to deal with the bg95 as 2 modems, one for connection and the other one for GPS, but I can't deal with it in the device tree.
I will be than thankful if you helped in this case. I opened a discussion about sending custom commands and triggering the script from the application layer. thanks in advance.

**Reply by bjarki-andreasen** [reply]
The plan is definitely to create two devices in devicetree, one for the cellular modem, implementing the experimental [cellular API ](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/cellular.h) the other implementing the [GNSS API](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/gnss.h).

These APIs (although handlers have not been created for syscalls yet) are meant to be called from user space/the application, it should not be necessary to send chat scripts manually from the application, the device driver APIs should take care of that :)

**Reply by AndrewAWG97** [reply]
thank you for your reply @bjarki-trackunit, I am using SDK 2.5.2 which doesn't have these GNSS API yet. so I want to try to make my own out-of-tree library. I just need a push to know how to launch my own script.

**Reply by bjarki-andreasen** [reply]
The modem module contexts are all allocateable in userspace, so you can create an instance of modem_chat, and a modem_backend_uart, and all of the scripts you want to in userspace, the only syscall required is from the modem_backend_uart to the UART device driver through the UART API, which is no problem :)

This example does just that (although its setting up CMUX and PPP as well) https://github.com/bjarki-trackunit/modem_modules/blob/main/samples/subsys/modem/cmux_ppp_tty/src/main.c though it is based on a very early version of the modem subsystem :)

**Reply by AndrewAWG97** [reply]
Thank you @bjarki-trackuni. The sample really helped me a lot to understand the sequence of the code. 
I want to run this code on my nRF52. I put my device in the device tree under uart0, but the code is not buildable; it needs the function static void ppp_iface_init(struct net_if *iface) to be implemented. So, with my current configuration how to implement this function?

**Reply by bjarki-andreasen** [reply]
Can you share the exact error produced by the compiler/linkter? I think you may have selected `CONFIG_NET_PPP=y` somewhere since this function is only used with the old `ppp.c` driver :)

**Reply by AndrewAWG97** [reply]
d:/cmux_ppp_tty/prj.conf:15: warning: attempt to assign the value 'y' to the undefined symbol NATIVE_POSIX_SLOWDOWN_TO_REAL_TIME. 
I can't build from the  beginning as this configuration is undefined.

**Reply by bjarki-andreasen** [reply]
This one has been changed to `CONFIG_NATIVE_SIM_SLOWDOWN_TO_REAL_TIME=y`

**Reply by AndrewAWG97** [reply]
@bjarki-trackunit 
/cmux_ppp_tty/prj.conf:15: warning: attempt to assign the value 'y' to the undefined symbol NATIVE_SIM_SLOWDOWN_TO_REAL_TIME
it's the same error. 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/116096087/bf7ad457-02a6-4d55-af45-fdf5293286bd)
I shared with you the proj.conf to make sure of it.

And below the build terminal. 

*  Executing task: nRF Connect: Build: cmux_ppp_tty/build (active) 

Building cmux_ppp_tty
west build --build-dir d:/Cypod/cmux_ppp_tty/build d:/Cypod/cmux_ppp_tty

[0/1] Re-running CMake...
Loading Zephyr default modules (Zephyr base (cached)).
-- Application: D:/Cypod/cmux_ppp_tty
-- CMake version: 3.20.5
-- Cache files will be written to: C:/ncs/v2.5.2/zephyr/.cache
-- Zephyr version: 3.4.99 (C:/ncs/v2.5.2/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.14.0")
-- Board: nrf52dk_nrf52832_CyLock
-- Found host-tools: zephyr 0.16.1 (C:/ncs/toolchains/c57af46cb7/opt/zephyr-sdk)
-- Found toolchain: zephyr 0.16.1 (C:/ncs/toolchains/c57af46cb7/opt/zephyr-sdk)
-- Found BOARD.dts: D:/Cypod/cylockv2i72/boards/arm/nrf52dk_nrf52832_CyLock/nrf52dk_nrf52832_CyLock.dts
-- Generated zephyr.dts: D:/Cypod/cmux_ppp_tty/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: D:/Cypod/cmux_ppp_tty/build/zephyr/include/generated/devicetree_generated.h
-- Including generated dts.cmake file: D:/Cypod/cmux_ppp_tty/build/zephyr/dts.cmake

warning: MODEM_BACKEND_TTY (defined at subsys/modem\backends/Kconfig:4) was assigned the value 'y'
but got the value 'n'. Check these unsatisfied dependencies: ARCH_POSIX (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_MODEM_BACKEND_TTY and/or look up
MODEM_BACKEND_TTY in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.


warning: The choice symbol NATIVE_UART_0_ON_STDINOUT (defined at
drivers/serial/Kconfig.native_posix:28) was selected (set =y), but no symbol ended up as the choice
selection. See http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_NATIVE_UART_0_ON_STDINOUT
and/or look up NATIVE_UART_0_ON_STDINOUT in the menuconfig/guiconfig interface. The Application
Development Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of
the manual might be helpful too.


Parsing C:/ncs/v2.5.2/zephyr/Kconfig
Loaded configuration 'D:/Cypod/cylockv2i72/boards/arm/nrf52dk_nrf52832_CyLock/nrf52dk_nrf52832_CyLock_defconfig'
Merged configuration 'd:/Cypod/cmux_ppp_tty/prj.conf'
warning: Experimental symbol NET_L2_PPP is enabled.


warning: Experimental symbol NET_CONNECTION_MANAGER is enabled.


d:/Cypod/cmux_ppp_tty/prj.conf:15: warning: attempt to assign the value 'y' to the undefined symbol NATIVE_SIM_SLOWDOWN_TO_REAL_TIME

error: Aborting due to Kconfig warnings

CMake Error at C:/ncs/v2.5.2/zephyr/cmake/modules/kconfig.cmake:348 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  C:/ncs/v2.5.2/nrf/cmake/modules/kconfig.cmake:29 (include)
  C:/ncs/v2.5.2/zephyr/cmake/modules/zephyr_default.cmake:129 (include)
  C:/ncs/v2.5.2/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  C:/ncs/v2.5.2/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:97 (include_boilerplate)
  CMakeLists.txt:6 (find_package)


-- Configuring incomplete, errors occurred!
See also "D:/Cypod/cmux_ppp_tty/build/CMakeFiles/CMakeOutput.log".
See also "D:/Cypod/cmux_ppp_tty/build/CMakeFiles/CMakeError.log".
FAILED: build.ninja 
C:\ncs\toolchains\c57af46cb7\opt\bin\cmake.exe --regenerate-during-build -SD:\Cypod\cmux_ppp_tty -BD:\Cypod\cmux_ppp_tty\build
ninja: error: rebuilding 'build.ninja': subcommand failed
FATAL ERROR: command exited with status 1: 'C:\ncs\toolchains\c57af46cb7\opt\bin\cmake.EXE' --build 'd:\Cypod\cmux_ppp_tty\build'

 *  The terminal process terminated with exit code: 1. 
 *  Terminal will be reused by tasks, press any key to close it.

**Comment by AndrewAWG97** [comment]
@bjarki-andreasen Hi bjarki, is there any way to change the cmds sent to the modem in runtime? 
for example if I want to change the mode from lte to gsm and vice versa ?

**Reply by bjarki-andreasen** [reply]
You can either implement the cellular.h API https://github.com/zephyrproject-rtos/zephyr/blob/e60da1bd640a37370870a83277142dd560f1fb8d/include/zephyr/drivers/cellular.h#L89-L91 and send the required cmds from the driver at runtime, or, if your modem supports it, you can use a user_pipe to send custom commands similar to how the at shell does it (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/modem/modem_at_shell.c)

**Reply by AndrewAWG97** [reply]
I can make custom scripts, thanks to your help and guidence, but I am trying to add use the command as an argument like this :

```
`char* at_cmd = "AT";

#if DT_HAS_COMPAT_STATUS_OKAY(quectel_eg25_g)
MODEM_CHAT_SCRIPT_CMDS_DEFINE(
	quectel_eg25_g_init_chat_script_cmds, MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
	MODEM_CHAT_SCRIPT_CMD_RESP("AT+CFUN=4", ok_match),
	MODEM_CHAT_SCRIPT_CMD_RESP(at_cmd, ok_match),`
```
	
C:/ncs/v2.7.0/zephyr/drivers/modem/modem_cellular.c:2441:9: note: in expansion of macro 'MODEM_CHAT_SCRIPT_CMD_RESP'
 2441 |         MODEM_CHAT_SCRIPT_CMD_RESP(at_cmd, ok_match),
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~
C:/ncs/v2.7.0/zephyr/include/zephyr/modem/chat.h:112:28: error: initializer element is not constant
  112 |                 .request = (uint8_t *)(_request), 
  
  I get this error.

**Reply by bjarki-andreasen** [reply]
change to `MODEM_CHAT_SCRIPT_CMD_RESP("AT", ok_match)` and remove at_cmd. If you need runtime configurable chat scripts, see https://github.com/zephyrproject-rtos/zephyr/blob/e60da1bd640a37370870a83277142dd560f1fb8d/drivers/gnss/gnss_quectel_lcx6g.c#L782-L798 :)

**Reply by AndrewAWG97** [reply]
I know, but there is no way to pass a string to the modem chat and change it in runtime? 
if the quectel_lcx6g_init_pair_script function do this, please explain more. Thanks in advance.

**Reply by mxc42** [reply]
@AndrewAWG97 were you ever able to figure this out? I'm struggling with the same thing.

**Reply by bjarki-andreasen** [reply]
There are two separate ways of creating chat scripts, const and runtime.

const chat scripts or not modifiable, they are defined using macros like `MODEM_CHAT_SCRIPT_CMDS_DEFINE()` these can not be changed at runtime, and they need to be passed const data, like char strings, not a pointer to a non-const buffer, which is not known at compile time, hence the `initializer element is not constant`

runtime chat scripts use the same structs naturally, but they are not defined as const like the macros do, they are defined as "normal" mutable structs. Look into the `struct quectel_lcx6g_data *data`, there you will find the "raw" chat scripts, matches and requests, as normal structs, which are then initialized and modified at runtime using the modem_chat runtime APIs like `modem_chat_script_chat_set_response_matches`

**Reply by mxc42** [reply]
Ok, I think I'm following. I'm using the `gnss_quectel_lcx6g` driver, if I want to send say a $PAIR690 message do I have to fork my own version of `gnss_quectel_lcx6g.c` and then somehow add my custom function to the `gnss_api`? I get the flow of lines 782-798 you sent earlier but I'm not sure how to add that to my own gnss file as `quectel_lcx6g_data` is defined in `gnss_quectel_lcx6g.c` and not in a header file.

If it is possible to get an instance of `data` in my own file, do I follow this format for sending the $PAIR690?:
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/gnss/gnss_quectel_lcx6g.c#L309-L350

**Reply by bjarki-andreasen** [reply]
You should copy the driver and keep it out-of-tree, then you can modify the data struct in your downstream version of the driver, and add a device driver specific header with the extra APIs you need out-of-tree as well.

**Reply by bjarki-andreasen** [reply]
custom device driver specific headers look like this https://github.com/zephyrproject-rtos/zephyr/pull/76087/commits/ecc3031d6dbd8d16f3e1e54314d2c15284d83d79#diff-555328e57afc6cacd9d4390c16e518d4808f5a81b7301e5b861060d94cba19cd

**Reply by mxc42** [reply]
Ok, I'll give that a go. Thanks!
